### PR TITLE
Fix: Lambda allow usage of partial ARN and ARN in function names

### PR DESCRIPTION
**Pull request by Morijarti**
## Motivation
From AWS documentation, function name can be
```

    Function name – my-function (name-only), my-function:v1 (with alias).
    Function ARN – arn:aws:lambda:us-west-2:123456789012:function:my-function .
    Partial ARN – 123456789012:function:my-function .
```

While `get_function` has support for this, `Update_function_configuration` does not.
This PR also aims to make sure that above function_name formats work in cross-account and cross region

## Changes
Modified `update_function_configuration` to extract function name, region and account from passed function_name instead of using it in raw format

## TODO
- [x] Verify that other lambda operations are following the same way of using function_name
- [x] run build with randomized regions and accounts https://app.circleci.com/pipelines/github/localstack/localstack/25711

### CODE REVIEW

**Comment by Morijarti** [comment]
While tests seem to be ok, I haven't added any new cross-account and cross-region tests, so please have a good look in case I missed something.
I can work on improving test coverage if needed.

**Comment by Morijarti** [comment]
I would suggest take a look at each usage of `context.account_id` in provider and see if we can replace it based on function ARN or partial ARN or something else.
For example at `event_source_mapping` methods

File: localstack-core/localstack/services/lambda_/provider.py [hunk 1]
```diff
@@ -973,12 +973,16 @@ def update_function_configuration(
         self, context: RequestContext, request: UpdateFunctionConfigurationRequest
     ) -> FunctionConfiguration:
         """updates the $LATEST version of the function"""
-        function_name = request.get("FunctionName")  # TODO: can this be an ARN too?
-        state = lambda_stores[context.account_id][context.region]
+        function_name = request.get("FunctionName")
+
+        # in case we got ARN or partial ARN
+        account_id, region = api_utils.get_account_and_region(function_name, context)
```
[the remark this answers is not in the document]
Comment (Morijarti) [reply localstack-core/localstack/services/lambda_/provider.py:979]: That makes total sense. I'll make the changes.

File: tests/aws/services/lambda_/test_lambda_api.py [hunk 2]
```diff
@@ -77,6 +77,37 @@ def environment_length_bytes(e: dict) -> int:
...
+
+        full_arn = create_response["CreateFunctionResponse"]["FunctionArn"]
+        partial_arn = ":".join(full_arn.split(":")[-3:])
+        valid_names = [full_arn, function_name, partial_arn]
+
+        # update configuration with various clarifiers
+        for name in valid_names:
+            aws_client.lambda_.update_function_configuration(
+                FunctionName=name,
+                Runtime=Runtime.python3_11,
```
[the remark this answers is not in the document]
Comment (Morijarti) [reply tests/aws/services/lambda_/test_lambda_api.py:102]: Should I remove these tests then?
I was mostly using them as debugging tool, but partial and full ARN weren;t covered in other tests

**Comment by joe4dev** [comment]
> While tests seem to be ok, I haven't added any new cross-account and cross-region tests, so please have a good look in case I missed something. I can work on improving test coverage if needed.

We won't be able to test all combinations, but it would should have at least one test for cross-region (this PR) and cross-account (follow-up). The most relevant operation(s) such as `GetFunction` should be fine. `GetLayerVersionByArn` is an interesting case as well because it interacts with the external layer fetcher (Pro feature).

Less important: We might want to watch out for cases where we accidentally enable cross-region or cross-account functionality despite not being supported at AWS 😆

Some interesting cases: I don't what happens if `CreateFunction` gets a ARN in a different region. Some context-specific operations (e.g., [ListFunctions](https://docs.aws.amazon.com/lambda/latest/api/API_ListFunctions.html)) have a dedicated `MasterRegion` marker 💡 , others such as `ListLayers` not.

File: tests/aws/services/lambda_/test_lambda_api.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (joe4dev) [reply tests/aws/services/lambda_/test_lambda_api.py:102]: The tests are good 👍 

> Or were you referring to usage of different runtime?
Yes, I think we can just omit the Runtime part :)

[the remark this answers is not in the document]
Comment (Morijarti) [reply tests/aws/services/lambda_/test_lambda_api.py]: Fixed

**Comment by Morijarti** [comment]
I've added simple cross region test.
Same test should work in cross account as well, but I need to setup IAM in AWS account in order for that.
Once I do that in future PR, I will rename the test.

File: tests/aws/services/lambda_/test_lambda_api.py [hunk 3]
```diff
@@ -77,6 +77,55 @@ def environment_length_bytes(e: dict) -> int:
...
+        function_name = f"fn-{short_uid()}"
+        create_response = create_lambda_function(
+            handler_file=TEST_LAMBDA_PYTHON_ECHO,
+            func_name=function_name,
+            runtime=Runtime.python3_12,
+            MemorySize=256,
+            Timeout=5,
+        )
+
+        aws_client.lambda_.get_waiter("function_active_v2").wait(FunctionName=function_name)
```
Comment (joe4dev) [inline tests/aws/services/lambda_/test_lambda_api.py:92]: FYI: The fixture does the waiting automatically, but I think it's a good idea to make it explicit here 👍

**Comment by joe4dev** [comment]
> Methods still using `context.account_id`. I'm unsure if we should change them as well
> 
> Function related:
> 
>     * create_function
> 
>     * list_functions
> 
> 
> Event source mapping:
> 
>     * create_event_source_mapping
> 
>     * update_event_source_mapping
> 
>     * delete_event_source_mapping
> 
>     * get_event_source_mapping
> 
>     * list_event_source_mapping
> 
> 
> Code signing config:
> 
>     * create_code_signing_config
> 
>     * update_code_signing_config
> 
>     * get_code_signing_config
> 
>     * get_function_code_signing_config
> 
>     * delete_code_signing_config
> 
>     * list_code_signing_configs
> 
>     * list_functions_by_code_signing_config
> 
> 
> Account concurrency and provision:
> 
>     * get_account_settings
> 
> 
> Layers:
> 
>     * publish_layer_version
> 
>     * list_layers

Great summary 👍 

I suggest:
* Refactor all these remaining `context` usages scattered within the code to use `region` and `account_id` initialized at the top of the function. They can still be initialized from the context, but it would make it more explicit and clear at first glance. Let me know if you think keeping the direct context usage is better.
* Compare this list against the [list of API actions that support resource-based policies](https://docs.aws.amazon.com/lambda/latest/dg/access-control-resource-based.html#permissions-resource-api). We currently assume that these are the ones supporting cross-account operations. Just flagging mismatches in this list would be helpful.
* The special case for [ListFunctions](https://docs.aws.amazon.com/lambda/latest/api/API_ListFunctions.html) could be tested in a follow-up using `MasterRegion`.

**Review by joe4dev — APPROVED** [review]
Bringing cross-region and cross-account compatibility to the next level 👍 

Two small suggestions before merging:
* Refactor scattered `context.*` usages (see previous reply comment)
* Trigger a build with randomized regions/accounts following our guide: https://github.com/localstack/localstack/blob/master/docs/testing/multi-account-region-testing/README.md
  * You can link this build from this PR
  * Highly recommended for such changes because they only run on master by default

**Pull request merged**